central/imagev2/datastore/mapper/datastore/datastore_impl.go

```diff
@@ -0,0 +1,208 @@
+	ds := &datastoreImpl{
+		imageDataStore:   datastoreV1,
+		imageV2DataStore: datastoreV2,
+
+		flattenImageData: features.FlattenImageData.Enabled(),
+	}
+	return ds
+}
+
+func (ds *datastoreImpl) SearchListImages(ctx context.Context, q *v1.Query) ([]*storage.ListImage, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.SearchListImages(ctx, q)
+	}
+
+	// Get v2 images
+	images, err := ds.imageV2DataStore.SearchRawImages(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert to list images
+	listImages := make([]*storage.ListImage, 0, len(images))
+	for _, image := range images {
+		// Convert v2 to v1
+		v1Image := mapper.ConvertToV1(image)
+		// Convert v1 to ListImage
+		listImage := types.ConvertImageToListImage(v1Image)
+		listImages = append(listImages, listImage)
+	}
+	return listImages, nil
+}
+
+func (ds *datastoreImpl) ListImage(ctx context.Context, sha string) (*storage.ListImage, bool, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.ListImage(ctx, sha)
+	}
+
+	// Get v2 image
+	image, found, err := ds.GetImageMetadata(ctx, sha)
+	if err != nil {
+		return nil, false, err
+	}
+	if !found {
+		return nil, false, nil
+	}
+	return types.ConvertImageToListImage(image), true, nil
+}
+
+func (ds *datastoreImpl) Search(ctx context.Context, q *v1.Query) ([]searchPkg.Result, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.Search(ctx, q)
+	}
+	return ds.imageV2DataStore.Search(ctx, q)
+}
+
+func (ds *datastoreImpl) Count(ctx context.Context, q *v1.Query) (int, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.Count(ctx, q)
+	}
+	return ds.imageV2DataStore.Count(ctx, q)
+}
+
+func (ds *datastoreImpl) SearchImages(ctx context.Context, q *v1.Query) ([]*v1.SearchResult, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.SearchImages(ctx, q)
+	}
+	return ds.imageV2DataStore.SearchImages(ctx, q)
+}
+
+func (ds *datastoreImpl) SearchRawImages(ctx context.Context, q *v1.Query) ([]*storage.Image, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.SearchRawImages(ctx, q)
+	}
+	images, err := ds.imageV2DataStore.SearchRawImages(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+	results := make([]*storage.Image, 0, len(images))
+	for _, image := range images {
+		results = append(results, mapper.ConvertToV1(image))
+	}
+	return results, nil
+}
+
+func (ds *datastoreImpl) CountImages(ctx context.Context) (int, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.CountImages(ctx)
+	}
+	return ds.imageV2DataStore.CountImages(ctx)
+}
+
+func (ds *datastoreImpl) GetImage(ctx context.Context, sha string) (*storage.Image, bool, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.GetImage(ctx, sha)
+	}
+	images, err := ds.imageV2DataStore.SearchRawImages(ctx, searchPkg.NewQueryBuilder().AddExactMatches(searchPkg.ImageSHA, sha).ProtoQuery())
```

Comment:
Can we check if it is a uuid and not a sha? If it is a uuid, then you can call `GetImage` and directly return that.